cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,221 @@
+      case 2:
+        F = B ^ C ^ D;
+        g = (3 * j + 5) % 16;
+        break;
+      case 3:
+        F = C ^ (B | (~D));
+        g = (7 * j) % 16;
+        break;
+    }
+
+    F = F + A + hash_constants[j] + buffer_ints[g];
+
+    A = D;
+    D = C;
+    C = B;
+    B = B + left_rotate(F, shift_constants[((j / 16) * 4) + (j % 4)]);
+  }
+
+  hash_state->hash_value[0] += A;
+  hash_state->hash_value[1] += B;
+  hash_state->hash_value[2] += C;
+  hash_state->hash_value[3] += D;
+
+  hash_state->buffer_length = 0;
+}
+
+template <typename Key>
+struct MD5Hash {
+  using argument_type = Key;
+
+  /**
+   * @brief Core MD5 element processing function
+   */
+  template <typename TKey>
+  void CUDA_HOST_DEVICE_CALLABLE process(TKey const& key,
+                                         const uint32_t len,
+                                         md5_intermediate_data* hash_state,
+                                         const md5_hash_constants_type* hash_constants,
+                                         const md5_shift_constants_type* shift_constants) const
+  {
+    uint8_t* data = (uint8_t*)&key;
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < 64) {
+      thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
+      hash_state->buffer_length += len;
+    } else {
+      uint32_t copylen = 64 - hash_state->buffer_length;
+
+      thrust::copy_n(thrust::seq, data, copylen, hash_state->buffer + hash_state->buffer_length);
+      md5_hash_step(hash_state, hash_constants, shift_constants);
+
+      while (len > 64 + copylen) {
+        thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+        md5_hash_step(hash_state, hash_constants, shift_constants);
+        copylen += 64;
+      }
+
+      thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  template <typename T, typename std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  void CUDA_HOST_DEVICE_CALLABLE operator()(T const& key,
+                                            md5_intermediate_data* hash_state,
+                                            const md5_hash_constants_type* hash_constants,
+                                            const md5_shift_constants_type* shift_constants) const
+  {
+    process(key, size_of(key), hash_state, hash_constants, shift_constants);
+  }
+
+  template <typename T, typename std::enable_if_t<!is_fixed_width<T>()>* = nullptr>
+  void CUDA_HOST_DEVICE_CALLABLE operator()(T const& key,
+                                            md5_intermediate_data* hash_state,
+                                            const md5_hash_constants_type* hash_constants,
+                                            const md5_shift_constants_type* shift_constants) const
+  {
+    CUDF_FAIL("Unsupported hash type");
+  }
+
+  void CUDA_HOST_DEVICE_CALLABLE operator()(Key const& key,
```

Comment:
Why is this one empty? Might warrant a comment if I'm not missing something trivial.

Reply:
I did a bunch of testing today to understand and explain this issue better. This `operatore()` is acting as a catch all case representing default behavior, that allows me to override specific cases like a string_view outside of the struct. Without this un-templated operator function, I get a bunch of compile errors `../include/cudf/detail/utilities/hash_functions.cuh(165): error: no suitable constructor exists to convert from "const __nv_bool" to "cudf::data_type"
` for a bunch of different types.

Ideally the default behavior should actually be a `CUDF_FAIL("Unsupported hash type")` but adding that to the empty function causes `../include/cudf/detail/utilities/hash_functions.cuh(173): error: device code does not support exception handling` errors.

During my testing this afternoon, non-fixed width types never hit the `CUDF_FAIL` on line 146 -- the same was seen for other types I was trying to filter out as unsupported column data types. It's clear the current method of filtering out unsupported types doesn't work, any guidance on how to fix this would be appreciated.

Reply:
Yeah, you cannot throw from device code. 
You can remove the specializations that you wanted top use to catch invalid types. Instead, do this in `md5_element_hasher`. It is executed on the host and thus can throw. Specialize its `operator()` to call CUDF_FAIL for unsupported types and you should be good with just valid implementations of `MD5Hash::operator()`.

Reply:
Yeah, this isn't right. You're not hitting `CUDF_FAIL` because you're doing device-side dispatch and calling your MD5Hash object in _device_ code. `CUDF_FAIL` is a host-only construct. I'm surprised this even compiles.

Reply:
You have to disable the types on the host side. Since both `MD5Hash` and `md5_element_hasher` are device code, you need a host-side dispatcher to filter the calls. Something like
https://github.com/rapidsai/cudf/blob/ad9dea1235d72674f99c6d626a45d4e1296d362d/cpp/src/copying/concatenate.cu#L284
That you call from type_dispatcher instead of directly calling `md5_element_hasher`. With this, you can remove this function and the one above with `CUDF_FAIL`.

Reply:
I'd suggest looking at `row_hasher` here: https://github.com/rapidsai/cudf/blob/855e735c07e14d11b8ef309d943e9a28c91e0698/cpp/include/cudf/table/row_operators.cuh#L411

Reply:
Required a bit of a rework, the use of `size_of` rather than `sizeof` was causing failures with primitive types and were being type dispatched to other operator functions.